pkg/repository/network.go

```diff
@@ -133,27 +276,427 @@ func (r *networkRepository) Create(ctx context.Context, rq *Validated[*apiv2.Net
 	}
 
 	for _, prefix := range nw.Prefixes {
-		_, err = r.r.ipam.CreatePrefix(ctx, connect.NewRequest(&ipamv1.CreatePrefixRequest{Cidr: prefix.String()}))
+		_, err = r.r.ipam.CreatePrefix(ctx, connect.NewRequest(&ipamv1.CreatePrefixRequest{Cidr: prefix.String(), Namespace: nw.Namespace}))
 		if err != nil {
 			return nil, err
 		}
 	}
-
 	return resp, nil
 }
 
-func (r *networkRepository) Update(ctx context.Context, msg *Validated[*apiv2.NetworkServiceUpdateRequest]) (*metal.Network, error) {
-	panic("unimplemented")
+func (r *networkRepository) Update(ctx context.Context, rq *Validated[*adminv2.NetworkServiceUpdateRequest]) (*metal.Network, error) {
+	old, err := r.Get(ctx, rq.message.Id)
+	if err != nil {
+		return nil, err
+	}
+	newNetwork := *old
+	req := rq.message
+
+	if req.Name != nil {
+		newNetwork.Name = *req.Name
+	}
+	if req.Description != nil {
+		newNetwork.Description = *req.Description
+	}
+	if req.Labels != nil {
+		for _, remove := range req.Labels.Remove {
+			delete(newNetwork.Labels, remove)
+		}
+		if req.Labels.Update != nil {
+			for k, v := range req.Labels.Update.Labels {
+				if newNetwork.Labels == nil {
+					newNetwork.Labels = make(map[string]string)
+				}
+				newNetwork.Labels[k] = v
+			}
+		}
+	}
+
+	if req.NatType != nil {
+		nt, err := metal.ToNATType(*req.NatType)
+		if err != nil {
+			return nil, err
+		}
+		newNetwork.NATType = &nt
+		switch nt {
+		case metal.IPv4MasqueradeNATType:
+			newNetwork.Nat = true
+		case metal.NoneNATType:
+			//
+		}
+	}
+
+	var (
+		prefixesToBeRemoved metal.Prefixes
+		prefixesToBeAdded   metal.Prefixes
+	)
+
+	prefixesToBeRemoved, prefixesToBeAdded, err = r.calculatePrefixDifferences(ctx, old, &newNetwork, req.Prefixes)
+	if err != nil {
+		return nil, errorutil.Convert(err)
+	}
+
+	if req.DestinationPrefixes != nil {
+		destPrefixes, err := metal.NewPrefixesFromCIDRs(req.DestinationPrefixes)
+		if err != nil {
+			return nil, errorutil.Convert(err)
+		}
+		newNetwork.DestinationPrefixes = destPrefixes
+	}
+
+	if req.DefaultChildPrefixLength != nil {
+		newDefaultChildPrefixLength, err := metal.ToChildPrefixLength(req.DefaultChildPrefixLength, newNetwork.Prefixes)
+		if err != nil {
+			return nil, errorutil.NewInvalidArgument(err)
+		}
+		newNetwork.DefaultChildPrefixLength = newDefaultChildPrefixLength
+	}
+
+	if req.Force {
```

Comment:
This should already be validated. Just checking whether in the request these are set and update if necessary is sufficient.